fedimint-server/src/net/api.rs

```diff
@@ -445,3 +534,35 @@ pub fn server_endpoints() -> Vec<ApiEndpoint<ConsensusApi>> {
         },
     ]
 }
+
+/// Very simple cache mostly used to protect endpoints against denial of service
+/// attacks
+#[derive(Clone)]
+pub struct ExpiringCache<T> {
+    data: Arc<tokio::sync::Mutex<Option<(T, Instant)>>>,
+    duration: std::time::Duration,
+}
+
+impl<T: Clone> ExpiringCache<T> {
+    pub fn new(duration: std::time::Duration) -> Self {
+        Self {
+            data: Arc::new(tokio::sync::Mutex::new(None)),
+            duration,
+        }
+    }
+
+    pub async fn get<Fut>(&self, f: impl FnOnce() -> Fut) -> T
+    where
+        Fut: futures::Future<Output = T>,
+    {
+        let mut data = self.data.lock().await;
```

Comment:
Since we're holding this lock over the call to `f()` there is a some room for deadlocks here, if the `f()` somehow triggered a recursive locking here. Which is unlikely, but not impossible as the caller might not realize it.

But I'm not sure if not holding it over `f()` is much better.

Reply:
If we don't hold the lock then multiple calls of `f()` may happen in parallel, defeating the whole purpose of this mechanism.

But if someone introduces a deadlock then it will probably happen on the first test/execution. It think it will be very clear that you did something wrong.

Reply:
I know. It would not help with a burst of messages, but it will still short-cut all subsequent messages (for 1s).

But yeah, probably not worth changing.